Implement RandomAccess.FlushToDisk()

ericmutta opened the pull request:
Fix #86836

ericmutta committed:
Implement FlushToDisk().

Fix #86836

danmoseley commented on src/libraries/System.Runtime/ref/System.Runtime.cs:
```diff
@@ -10023,6 +10023,7 @@ public static partial class RandomAccess
         public static void Write(Microsoft.Win32.SafeHandles.SafeFileHandle handle, System.ReadOnlySpan<byte> buffer, long fileOffset) { }
         public static System.Threading.Tasks.ValueTask WriteAsync(Microsoft.Win32.SafeHandles.SafeFileHandle handle, System.Collections.Generic.IReadOnlyList<System.ReadOnlyMemory<byte>> buffers, long fileOffset, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
         public static System.Threading.Tasks.ValueTask WriteAsync(Microsoft.Win32.SafeHandles.SafeFileHandle handle, System.ReadOnlyMemory<byte> buffer, long fileOffset, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        public static void FlushToDisk(Microsoft.Win32.SafeHandles.SafeFileHandle handle) { throw null; }
```
nit - please put in alpha order. we have tools that regenerate this and it will otherwise cause a diff

ericmutta replied:
OK...before I do, I notice that `SetLength` appears before the various `Read` and `ReadAsync` methods, when it should appear after them in order to preserve alpha order. Should I make that change in this PR too or is it best to use a separate issue?

danmoseley replied:
If you run the tool it will take care of it

https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md

ericmutta committed:
Fix typo.

This "platform-depedent" should read "platform-dependent".

Co-authored-by: Dan Moseley <danmose@microsoft.com>

ericmutta committed:
Address PR feedback.

Declarations of class members in System.Runtime.cs should be in
alphabetic order (see https://github.com/dotnet/runtime/pull/89100#discussion_r1267026133).

Simplify comments on the call to "FileStreamHelpers.FlushToDisk()"
(see https://github.com/dotnet/runtime/pull/89100#discussion_r1267030174).

ericmutta commented on src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.cs:
```diff
@@ -250,7 +251,46 @@ public static ValueTask WriteAsync(SafeFileHandle handle, IReadOnlyList<ReadOnly
+            ValidateInput(handle, fileOffset: 0, flushingToDisk: true);
+
+            // Call the same helper function that is used when users call FileStream.Flush(flushToDisk: true).
+            // While this was done to save time, it has the added bonus of providing consistent behavior which is
```
I have removed the "save time" comment. That was written from my perspective when I was pleasantly surprised to find that there's an existing helper method with the exact same name and purpose as the new `RandomAccess.FlushToDisk()` method that I was implementing 👍

adamsitnik commented on src/libraries/System.IO.FileSystem/tests/RandomAccess/FlushToDisk.cs:
```diff
@@ -0,0 +1,103 @@
+                // Get the file time AFTER flushing to disk.
+                DateTime fileTimeAfterFlush = File.GetLastWriteTimeUtc(testFilePath);
+
+                // After explicitly flushing to disk, we expect "fileTimeAfterFlush > fileTimeBeforeFlush".
```
thank you for thinking about edge cases up-front 👍

adamsitnik commented on src/libraries/System.IO.FileSystem/tests/RandomAccess/NoBuffering.Windows.cs:
```diff
@@ -242,6 +242,80 @@ public async Task ReadShouldReturnZeroForEndOfFile(bool asyncOperation, bool asy
+            string testFilePath = GetTestFilePath();
+
+            // Generate random bytes to write to file. NOTE: since we use unbuffered I/O,
+            // the byte count must be a multiple of the storage device's sector size. We
```
this is a very valuable comment, thank you for adding it 👍

adamsitnik commented on src/libraries/System.IO.FileSystem/tests/RandomAccess/NoBuffering.Windows.cs:
```diff
@@ -242,6 +242,80 @@ public async Task ReadShouldReturnZeroForEndOfFile(bool asyncOperation, bool asy
+            // powers of 2 meaning the system page size (which is typically 4,096 bytes) will
+            // be a multiple of the sector size (which is typically 512 bytes) hence meeting
+            // the requirements for unbuffered I/O.
+            int byteCount = Random.Shared.Next(1, 10) * Environment.SystemPageSize;
```
similarly to other test, I would prefer to make it a theory and provide multiple different values in explicit way as arguments (test cases)

ericmutta replied:
Fixed with [this commit](https://github.com/dotnet/runtime/pull/89100/commits/276f2c1706f60b26882c87acc2265c45bd8c7a14) 👍

adamsitnik reviewed (commented):
@ericmutta great job, thank you for your contribution! PTAL at my comments

ericmutta commented:
Many thanks @adamsitnik for the detailed and encouraging feedback! I will do the necessary over the next 48 hours and push more commits 👍

danmoseley commented on src/libraries/System.IO.FileSystem/tests/RandomAccess/FlushToDisk.cs:
```diff
@@ -0,0 +1,103 @@
+            // Save test file path so we can refer to the same file throughout the test.
+            string testFilePath = GetTestFilePath();
+
+            // Generate random bytes to write to file. To ensure that flushing works correctly for a variety of
+            // sizes, we want to test with buffers that are smaller than a page (e.g. just 1 byte) and buffers
+            // that are several times larger than a page (e.g. up to 10 pages).
```
agreed, and you will probably need to use `[MemberData]` since you need to evaluate `Environment.SystemPageSize` to generate the list of variations for xunit.

example: https://github.com/dotnet/runtime/blob/d512f19ef3a506c01bb3a8588de48cb377472639/src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.Match.Tests.cs#L1036

danmoseley commented on src/libraries/System.IO.FileSystem/tests/RandomAccess/FlushToDisk.cs:
```diff
@@ -0,0 +1,103 @@
+                RandomAccess.Write(handle, randomBytes, fileOffset: 0);
+
+                // Get the file time BEFORE flushing to disk.
+                DateTime fileTimeBeforeFlush = File.GetLastWriteTimeUtc(testFilePath);
```
consider setting a last write time here that's a little older. otherwise you'll fail when the file system timestamp granularity isn't high enough. at an extreme, this test will fail on a FAT32 partition.

also, this test will fail if a daylight savings change or leap second happens during the test. obviously, that's unlikely, but several of our tests fail every time there's such an event as there's zillions of tests and they run all the time. to fix that, you can set a date that's a year ago, for example.

ericmutta replied:
Interesting points @danmoseley! In this test we are not explicitly setting the last write time so I can't control it in any way. When the file is flushed, **it is the operating system that updates the last write time** and the test is just checking that the time after flushing is greater than the time before flushing (or that they are at least equal, as explained in the comment near the `Assert` at the end).

The most important comment in the test says:

```csharp
 // Flush the file to disk. As a bare minimum test, this should work without throwing an
 // exception on all supported platforms. 
```

...i.e. we are mostly interested in the fact that flushing works without throwing an exception on all platforms. So I can remove all the date-related asserts to avoid the issues you raised. Please let me know if I should do this 👍

danmoseley commented on src/libraries/System.IO.FileSystem/tests/RandomAccess/FlushToDisk.cs:
```diff
@@ -0,0 +1,103 @@
+                // to. This "implicit flush" would also update the file time and if it happens BEFORE we set the
+                // fileTimeBeforeFlush variable above, then both fileTimeBeforeFlush and fileTimeAfterFlush will
+                // end up with the same value, which is why we use the ">=" operator below instead of using ">".
+                Assert.True(fileTimeAfterFlush >= fileTimeBeforeFlush);
```
another reason to set your own older timestamp before the flush is you eliminate this race. and then you can use Assert.NotEqual here instead. which also has the advantage that if it fails, it will be clear why, whereas when you use Assert.True it just says false expected true.

ericmutta replied:
Please see [this response](https://github.com/dotnet/runtime/pull/89100#discussion_r1273489917), I believe the decision we make there will address this too 👍

danmoseley commented on src/libraries/System.IO.FileSystem/tests/RandomAccess/NoBuffering.Windows.cs:
```diff
@@ -242,6 +242,80 @@ public async Task ReadShouldReturnZeroForEndOfFile(bool asyncOperation, bool asy
+        [Fact]
+        public void ReadFromDiskAfterFlushToDisk()
+        {
+            // Save test file path so we can refer to the same file throughout the test.
```
```suggestion
```
keep comments to non obvious :)

ericmutta replied:
I hear you :)

I once worked on a project where every line of code was commented, even the tests! Coming from the "zero comments unless absolutely necessary" camp, I found this practice really perplexing so I asked the team leader why everyone on the team did this. **I remember his response word for word because it profoundly changed the way I write code**. He said:

> Document everything. If it is not already obvious to the reader, then the reader will gain understanding. If it is already obvious to the reader, then the reader will be reassured that their understanding is correct.

So to this day I generally follow the "document everything" approach to commenting and it has been made even easier to do by GitHub CoPilot which will happily write much of the code for me when I document even the obvious stuff.

In any case, I have followed your recommendation and removed many of the non-essential comments in [this commit](https://github.com/dotnet/runtime/pull/89100/commits/a7d6289990c02f84d6d2b79fbd7959ded2ceee74). Some comments remain and I hope they add value but if not, please let me know and I will be happy to remove all of them 👍

danmoseley replied:
Yeah, it's a matter of taste to some extent. Too many comments can make code less readable just like too few comments does. Good naming, structure etc can sometimes do the job.

ericmutta committed:
Trim non-essential comments.

This includes comments that:

1. Can become outdated if methods are renamed (see https://github.com/dotnet/runtime/pull/89100#discussion_r1267742108).

2. Are not absolutely critical (see https://github.com/dotnet/runtime/pull/89100#discussion_r1268482921).

I also removed the mention of using FileOptions.WriteThrough as an
alternative to RandomAccess.FlushToDisk(). This is an advanced scenario
that can actually degrade performance if done incorrectly (i.e. if the
user doesn't manage their own buffers such that all writes keep hitting
the disk) so it's probably not wise to mention it unless we go into more
detail than is ideal for inclusion in the Remarks section of an XML doc
comment.

To remove the lengthy comment in the ValidateInput() function I had to
rename the optional parameter from "flushingToDisk" to the more generic
and self-describing "allowUnseekableHandles" (see https://github.com/dotnet/runtime/pull/89100#discussion_r1267724547).

ericmutta commented on src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.cs:
```diff
@@ -250,7 +251,44 @@ public static ValueTask WriteAsync(SafeFileHandle handle, IReadOnlyList<ReadOnly
+        /// you need to open the file using the <see cref="FileOptions.WriteThrough"/> option which causes all writes to the file to
+        /// be flushed to disk immediately such that you no longer need to call this method. In that scenario, you would allocate and
+        /// manage your own buffers then call <see cref="WriteAsync(SafeFileHandle, ReadOnlyMemory{byte}, long, CancellationToken)"/>
+        /// to flush a specific buffer to disk.
```
I ended up removing the last paragraph entirely in [this commit](https://github.com/dotnet/runtime/pull/89100/commits/a7d6289990c02f84d6d2b79fbd7959ded2ceee74) (see the commit message for the rationale).

ericmutta commented on src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.cs:
```diff
@@ -250,7 +251,44 @@ public static ValueTask WriteAsync(SafeFileHandle handle, IReadOnlyList<ReadOnly
+        /// This method calls platform-dependent APIs such as <c>FlushFileBuffers()</c> on Windows and <c>fsync()</c> on Unix.
+        /// </para>
+        /// <para>
+        /// Flushing the buffers causes data to be written to disk which is a relatively expensive operation. It is recommended
```
Thanks @adamsitnik! I was a bit nervous about writing such detailed comments (since the default for much of the libraries is to have no comments at all) but I am glad some of the comments have proven useful 🙏

ericmutta commented on src/libraries/System.IO.FileSystem/tests/RandomAccess/FlushToDisk.cs:
```diff
@@ -0,0 +1,103 @@
+    public partial class RandomAccess_FlushToDisk : RandomAccess_Base<long>
+    {
+        // Setting this to false disables the ThrowsArgumentOutOfRangeExceptionForNegativeFileOffset() test
+        // which is not applicable to FlushToDisk() since FlushToDisk() does not take a file offset parameter.
```
You are correct. I have fixed this in [this commit](https://github.com/dotnet/runtime/pull/89100/commits/a7d6289990c02f84d6d2b79fbd7959ded2ceee74) by removing the comment entirely, along with other non-essential comments in order to avoid issues like that 👍

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.cs:
```diff
@@ -306,14 +297,7 @@ private static void ValidateInput(SafeFileHandle handle, long fileOffset, bool f
                     ThrowHelper.ThrowObjectDisposedException_FileClosed();
                 }
 
-                // NOTE: this function was unconditionally throwing an exception for all non-seekable handles but
```
this comment was valuable, please just move it to `FlushToDisk(SafeFileHandle handle)` rather than deleting

ericmutta replied:
Done via [this commit](https://github.com/dotnet/runtime/pull/89100/commits/4748b8a41da7235973b159db6adcc023c880fb39) 👍

ericmutta committed:
Restore important comment.

The differences between Unix and Windows when it comes to flushing
unseekable handles is worth noting explicitly and the comment that
did this was previously deleted but has now been restored with some
minor edits to fit its new context (see https://github.com/dotnet/runtime/pull/89100/commits/a7d6289990c02f84d6d2b79fbd7959ded2ceee74#r1270505888).

ericmutta committed:
Convert some tests to theories.

See the following discussions for the rationale:
  - https://github.com/dotnet/runtime/pull/89100#discussion_r1267745260
  - https://github.com/dotnet/runtime/pull/89100#discussion_r1267751129

NOTE: the [MemberData] used by the theories returns one element in order
to reduce the number of tests that need to run per build. Since the element
is random, different builds will likely pick a different value allowing
more variations to be tested over many build cycles.

ericmutta commented:
Thanks @danmoseley and @adamsitnik for all the feedback on this! I have addressed most of the issues and will be back to resolve the rest over the next few days (**contributing to .NET is dangerously addictive, I wish I could do it full time**...but until then I have to restrain myself and attend to the day job before I get fired 😄 🏃 )

danmoseley commented on src/libraries/System.IO.FileSystem/tests/RandomAccess/FlushToDisk.cs:
```diff
@@ -0,0 +1,99 @@
+            // To ensure that flushing works correctly for a variety of sizes, we want to test with
+            // buffers that are smaller than a page (e.g. just 1 byte) and buffers that are several
+            // times larger than a page (e.g. up to 10 pages).
+            new object[] { Random.Shared.Next(1, Environment.SystemPageSize * 10) },
```
this needs to use a local Random object that was seeded. Eg., you see `new Random(42)` across many tests. Otherwise failures aren't reproducible.

danmoseley replied:
for something like the file content, it doesn't matter, as we don't think that should be relevant.

danmoseley replied:
Or instead of random it may be better to just list some fixed values eg 1, half page size, page size, etc

ericmutta replied:
Change made in [this commit](https://github.com/dotnet/runtime/pull/89100/commits/54a321514f93e2904f4ccaec28208ef8c82fddfe) 👍

adamsitnik reviewed (commented):
It looks very good, we are almost there 👍 

Again thank you for your contribution @ericmutta !

ericmutta committed:
Avoid randomness in ReadFromDiskAfterFlushToDisk().

See https://github.com/dotnet/runtime/pull/89100#discussion_r1271930026

ericmutta committed:
Avoid randomness in UpdatesFileLastWriteTime().

See the following:
 - https://github.com/dotnet/runtime/pull/89100#discussion_r1271056788
 - https://github.com/dotnet/runtime/pull/89100#discussion_r1271926007

ericmutta commented on src/libraries/System.IO.FileSystem/tests/RandomAccess/NoBuffering.Windows.cs:
```diff
@@ -242,6 +254,62 @@ public async Task ReadShouldReturnZeroForEndOfFile(bool asyncOperation, bool asy
+        {
+            // Sanity check: we expect the buffer byte count to be a multiple of the system page size
+            // since we use that as a proxy for the storage sector size.
+            Assert.Equal(0, byteCountForUnbufferedIO % Environment.SystemPageSize);
```
Change made in [this commit](https://github.com/dotnet/runtime/pull/89100/commits/8ab5bee834972ba9a96187ed80f00d0f749a5c23) 👍

ericmutta commented on src/libraries/System.IO.FileSystem/tests/RandomAccess/FlushToDisk.Unix.cs:
```diff
@@ -0,0 +1,38 @@
+            // Open the file for reading.
+            using (SafeFileHandle handle = File.OpenHandle(testFilePath, FileMode.Open, FileAccess.Read))
+            {
+                // On non-Windows platforms (notably Unix), flushing a file opened for reading should succeed.
+                // It appears that these platforms initially behaved like Windows but then changed their
+                // behavior (see https://www.austingroupbugs.net/view.php?id=671 for the discussion and
+                // https://pubs.opengroup.org/onlinepubs/9699919799/functions/aio_fsync.html for the spec
+                // containing the words "Note that even if the file descriptor is not open for writing...").
+                RandomAccess.FlushToDisk(handle);
```
I agree with you @adamsitnik that aligning the behaviours will make life easier for users of this function 👍 

Currently the XML docs I added contain this:

```csharp
        /// <exception cref="T:System.UnauthorizedAccessException">
        /// <paramref name="handle" /> was not opened for writing (on Windows). <-----
        /// </exception>
```

...i.e. they state that if you flush a file opened for reading (for whatever reason), the `UnauthorizedAccessException` is only thrown on Windows. We have three choices here:

1. **Leave it as it is right now:** this means the behaviour is different between platforms but it allows users on Unix to flush files opened for reading. While I can't think of a particular use case for this, it seems to have been important enough to some people to warrant a change in the POSIX spec as per [this discussion](https://www.austingroupbugs.net/view.php?id=671).
2.  **Make Windows behave like Unix:** this means we would catch the `UnauthorizedAccessException` on Windows and ignore it so no error is reported when you flush a file opened for reading on all platforms. This would allow the Unix users to keep the feature and is similar to how we currently ignore errors on Unix when someone flushes a non-seekable file, so we can allow the Windows users to keep that feature in order to flush pipe handles.
3. **Make Unix behave like Windows:** this means we would manually check if a file is opened for reading, then throw `UnauthorizedAccessException` so flushing such a file throws the same error on both platforms. This does give us consistency but it prevents Unix users from doing whatever it is that people had in mind when they [changed the spec](https://www.austingroupbugs.net/view.php?id=671) to allow flushing files opened for reading.

I am still undecided myself but I am leaning towards number 2: **make Windows behave like Unix** which means swallowing errors on Windows so flushing a file opened for reading does NOT throw on either platform. That gives us the consistency we want, while still allowing Unix users to take advantage of that feature (even though we may not be sure what the use case is).

@danmoseley any thoughts here?

ericmutta replied:
@adamsitnik thanks for the code snippet! I have made the suggested changed in [this commit](https://github.com/dotnet/runtime/pull/89100/commits/749d9b83d7318521b360fba0a4e2a9e2a8a0fe44) 👍

adamsitnik reviewed (commented):
@ericmutta apologies for the delay

ericmutta committed:
Harmonize flushing of read-only files.

There is a difference in behaviour between FlushFileBuffers() on
Windows and fsync() on Unix when it comes to read-only files. On
Windows it fails with ERROR_ACCESS_DENIED while on Unix it just works.
We made a decision to harmonize the platforms by making Windows behave
like Unix (i.e. we catch and ignore the error if a user attempts to
flush a file opened for reading, so it is effectively a no-op).

See https://github.com/dotnet/runtime/pull/89100#discussion_r1283070967

adamsitnik reviewed (approved):
LGTM, thank you for your contribution @ericmutta !

adamsitnik commented:
@jeffhandley I request for the permission to merge. This PR adds `FlushToDisk` method to `RandomAccess` class. It was already present for years on `FileStream`, it reuses an existing helper. So the risk for breaking anything else is very small.

jeffhandley reviewed (approved):
I approve of merging this in for .NET 8 RC1.

It adds a targeted new API that was approved a couple months ago, reusing existing code to implement the logic, and there's sufficient testing added.

adamsitnik commented:
The failure is unrelated (https://github.com/dotnet/runtime/issues/87879), merging!

danmoseley commented:
Thanks @ericmutta ! We hope you will become a regular contributor here.

ericmutta commented:
Thanks @danmoseley!

I hope to be back after we ship .NET 8 later this year, to contribute to .NET 9.

As part of the database-like project that led me to request `RandomAccess.FlushToDisk()`, I have had to build some things from scratch that I wish were already in the .NET libraries. Currently, the `SemaphoreSlim` class is the only synchronization primitive in the `System.Threading`   namespace that has some async support. If there's any interest from the .NET team, I would like to take a stab at adding more async support to other primitives like `ReaderWriterLockSlim` during the .NET 9 time frame next year 👍

adamsitnik commented:
> Currently, the SemaphoreSlim class is the only synchronization primitive in the System.Threading namespace that has some async support. If there's any interest from the .NET team, I would like to take a stab at adding more async support to other primitives like ReaderWriterLockSlim during the .NET 9 time frame next year 

cc @stephentoub

danmoseley commented:
>  If there's any interest from the .NET team, I would like to take a stab at adding more async support to other primitives like ReaderWriterLockSlim during the .NET 9 time frame next year 👍

the best way to see whether there's interest is to start an API proposal. Be sure to search existing issues though, stuff like this has likely been proposed before. (in which case, if it wasn't rejected, it may just need someone to take an interest)
